### DISCUSSION TITLE
Support to work with OpenAudible

### DISCUSSION THREAD

**Discussion by monkeyman512** [advplyr/audiobookshelf#401] [Ideas]
It would be nice if I could setup the [OpenAudible ](https://openaudible.org/) app to backup my Audible purchases then have the files incorporated as a library.

Edit:
To be clear. I want to point audiobookshelf at the "Library Directory" created by OpenAudible and have it recognize all the books.

**Comment by openaudible** [comment]
Great idea!

OpenAudible defaults to "1 book per file" with all audiobooks in the root directory. 

What if Audiobook Shelf's policy that "Files in the root directory are ignored" was simply removed?

New logic: If there is an audio file at the root level, it is assumed to be a full self-contained audiobook.

This change should get 95% compatibility with a minimum amount of changes.  

Another option would be to look for our books.json file, and if found, read that in and skip the ffprobe parsing, where you might pick up additional meta data that might be useful-- links to audible for the book, author or series, full chapter info, date purchased, etc. 

Ping us if anyone wants an OpenAudible license for development or testing or wants to collaborate on this.

**Reply by advplyr** [reply]
We can definitely remove the restriction of ignoring files in the root directory. That was put in early in development before metadata was probed and the scanner relied on folder names.
Is there any valuable metadata in the book.json file that audiobookshelf can use as well?

**Reply by advplyr** [reply]
@MediaCowboy was kind enough to share all the details about the folder structure and books.json file.
We will add support for mapping the OpenAudible root folder in which case it will see a `books.json` file and know to look in the `books` folder.
OR
Mapping the `books` folder in which case each audio file will be treated as a book.

**Reply by mediacowboy** [reply]
Glad I could be of help. Both of these programs are awesome

**Reply by openaudible** [reply]
That's great news about audiobook files in the root directory. That might just do the trick for a quick fix. 

Our users have been asking to be able to organize the file structure for a long time, so that is something we're about to let happen.. which might break things later if they save books as {author}/{title}.m4b. Which might get imported on your side as a book named author with chapters for each book.. so at some point we both would need to figure out the best way to mark a directory as either a a single book with chapters, or a directory with multiple books. My thought was to just do a "best guess" and let the user mark it as a single book if we guess wrong. 

Here is what we have for json items in the books.json file.. although some might be used very infrequently or not at all.
 
```
    product_id, asin, info_link, title, author, narrated_by, summary, description, duration, rating_average,
    rating_count, release_date, purchase_date, publisher, genre, sub_genre, title_short, copyright,
    author_link, download_link, filename, key, series_link, series_name, series_sequence, info_link_error,
    abridged, language, ignored, pdf_url, image_url, dir_path, region, chapters, locked, ayce, tags, notes
```
Mostly you might be interested in date_purchased,  the URL links, and series info. 

Chapters are a json array of 
```
    "start_offset_ms": 27368895,
    "length_ms": 891368,
    "title": "Epilogue",
    "start_offset_sec": 27369
```

We also encode the json for the whole book and save it as a meta data tag in the audiobook (mp3/m4b) file.. so if ffprobe finds a tag named "json64", you can base64.decode it and becomes a json object with data that wouldn't normally be something that can be tagged.  

And currently, the books.json is in the user's prefs directory, and the location of the actual library is found in a directories.json file. So a little confusing-- and the result of building it and not wanting to break it for existing users. So that may change to support multiple libraries, etc. 

But would be great to be able to offer some options for users who want a better web experience. We offer a very basic html output that has no player.. so yours seems like a great improvement.

**Reply by BCNelson** [reply]
Did this go anywhere? I'm not seeing this functionality and taking a quick look at the code and didn't see anything.

**Reply by advplyr** [reply]
Since the folder structure for openaudible may be changing I did not add additional checks for books.json.
What was added as a result of this is if you have audio files in the root of your library folder they will be scanned in as individual audiobooks.
